### PR TITLE
Password check

### PR DESCRIPTION
**Pull request by ms10335**

### CODE REVIEW

**Comment by ziobron** [comment]
✅ CI Passed - Task: password-check
🏅 Granted 8 XP for @ms10335

**Comment by ziobron** [comment]
⏱ 3 XP extra for @ms10335 for on-time delivery

**Comment by ziobron** [comment]
✅ CI Passed - Task: password-check-bonus
🏅 Granted 4 XP for @ms10335

File: homework/password-check/validation.cpp [hunk 1]
```diff
@@ -1,2 +1,82 @@
...
+std::map<ErrorCode, std::string> convertEnumsToString() {
+    std::map<ErrorCode, std::string> result{
+        {ErrorCode::Ok, "Ok"},
+        {ErrorCode::PasswordNeedsAtLeastNineCharacters, "Password needs to have at least nine characters"},
+        {ErrorCode::PasswordNeedsAtLeastOneNumber, "Password needs to have at least one number"},
+        {ErrorCode::PasswordNeedsAtLeastOneSpecialCharacter, "Password needs to have at least one special character"},
+        {ErrorCode::PasswordNeedsAtLeastOneUppercaseLetter, "Password needs to have at least one uppercase letter"},
+        {ErrorCode::PasswordsDoNotMatch, "Passwords do not match"}};
+    return result;
+}
```
Comment (nauka-programowania-MA) [inline homework/password-check/validation.cpp:17]: Should be const standalone function

File: homework/password-check/validation.cpp [hunk 2]
```diff
@@ -1,2 +1,82 @@
...
+std::string getErrorMessage(ErrorCode er) {
+    auto res = convertEnumsToString();
+    return res.at(er);
+}
+bool doPasswordsMatch(std::string password1, std::string password2) {
```
Comment (nauka-programowania-MA) [inline homework/password-check/validation.cpp:22]: Add space between functions

File: homework/password-check/validation.cpp [hunk 3]
```diff
@@ -1,2 +1,82 @@
...
+    if (password1.size() != password2.size()) {
+        return false;
+    }
+    if ((password1.empty() == password2.empty())) {
```
Comment (nauka-programowania-MA) [inline homework/password-check/validation.cpp:26]: I guess empty password is wrong password ;p

File: homework/password-check/validation.cpp [hunk 4]
```diff
@@ -1,2 +1,82 @@
...
+        return true;
+    } 
+    else if (password1.size() == password2.size()) {
```
Comment (nauka-programowania-MA) [inline homework/password-check/validation.cpp:29]: When you use return inside `if` you don't need `if else` or `else`. Just write `if (sth) { return blabla; }`

File: homework/password-check/validation.cpp [hunk 5]
```diff
@@ -1,2 +1,82 @@
...
+        auto pair = std::mismatch(password1.begin(), password1.end(), password2.begin());
+        if (pair.first == password1.begin()) {
+            return true;
+        }
+    }
```
Comment (nauka-programowania-MA) [inline homework/password-check/validation.cpp:34]: Just return lsh == rhs

File: homework/password-check/validation.cpp [hunk 6]
```diff
@@ -1,2 +1,82 @@
...
+    return false;
+}
+ErrorCode checkPasswordRules1(std::string password) {
```
Comment (nauka-programowania-MA) [inline homework/password-check/validation.cpp:37]: Pass by const refrence. Try always avoid copy.

File: homework/password-check/validation.cpp [hunk 7]
```diff
@@ -1,2 +1,82 @@
...
+    std::srand(time(nullptr));
+    auto res = convertEnumsToString();
+    int rand = std::rand() % (res.size());
+    std::map<ErrorCode, std::string>::iterator it = res.begin();
+    int i = 0;
+    while (i < rand) {
+        ++it;
+        ++i;
+    }
+    return it->first;
```
Comment (nauka-programowania-MA) [inline homework/password-check/validation.cpp:47]: Why so compilated?

File: homework/password-check/validation.cpp [hunk 8]
```diff
@@ -1,2 +1,82 @@
...
+}
+ErrorCode checkPasswordRules(std::string password) {
```
Comment (nauka-programowania-MA) [inline homework/password-check/validation.cpp:49]: Pass by const refrence. Try always avoid copy.

File: homework/password-check/validation.cpp [hunk 9]
```diff
@@ -1,2 +1,82 @@
...
+
+
+    if(password.size() < 9){
```
Comment (nauka-programowania-MA) [inline homework/password-check/validation.cpp:52]: Add space after `if`, `for`, `while` `function` etc... and also after `... < 9) {`

File: homework/password-check/validation.cpp [hunk 10]
```diff
@@ -1,2 +1,82 @@
...
+    const std::regex Capital_letter("[A-Z]+");
+    std::smatch pass_match, capital_match;
+    if(!(std::regex_search(password,pass_match,special_char_pattern)))
+    {
+        return ErrorCode::PasswordNeedsAtLeastOneSpecialCharacter;
+    }
+    if(!(std::regex_search(password,capital_match,Capital_letter)))
+    {
+        return ErrorCode::PasswordNeedsAtLeastOneUppercaseLetter;
+    }
```
Comment (nauka-programowania-MA) [inline homework/password-check/validation.cpp:71]: Just use non_of with 3 preduciate, is_digit, is_num, is_alphanum

File: homework/password-check/validation.cpp [hunk 11]
```diff
@@ -1,2 +1,82 @@
...
+    //if all valid is passed
+    return ErrorCode::Ok;
+}
+ErrorCode checkPassword(std::string password1, std::string password2) {
```
Comment (nauka-programowania-MA) [inline homework/password-check/validation.cpp:75]: Pass by const refrence. Try always avoid copy.

File: homework/password-check/validation.cpp [hunk 12]
```diff
@@ -1,2 +1,82 @@
...
+    auto res = convertEnumsToString();
+    if (!doPasswordsMatch(password1, password2)) {
+        return ErrorCode::PasswordsDoNotMatch;
+    } else {
+        return checkPasswordRules(password1);
```
Comment (nauka-programowania-MA) [inline homework/password-check/validation.cpp:80]: When you use return inside `if` you don't need `if else` or `else`. Just write `if (sth) { return blabla; }`

File: homework/password-check/validation.cpp [hunk 13]
```diff
@@ -1,2 +1,82 @@
...
+    }
+    if(!(std::regex_search(password,capital_match,Capital_letter)))
+    {
+        return ErrorCode::PasswordNeedsAtLeastOneUppercaseLetter;
+    }
+    //if all valid is passed
+    return ErrorCode::Ok;
+}
+ErrorCode checkPassword(std::string password1, std::string password2) {
+    auto res = convertEnumsToString();
```
Comment (nauka-programowania-MA) [inline homework/password-check/validation.cpp:76]: should be const

**Pull request closed without merging**